<div style="background: linear-gradient(135deg, #1a1a2e 0%, #16213e 50%, #0f3460 100%); padding: 60px 40px; border-radius: 16px; text-align: center; font-family: 'Segoe UI', sans-serif;">

<div style="font-size: 52px; margin-bottom: 10px;">✈️</div>

<h1 style="color: #e94560; font-size: 36px; margin: 0 0 8px 0; letter-spacing: 2px;">Flight Delay Analysis</h1>
<h2 style="color: #a8b2d8; font-size: 18px; font-weight: 400; margin: 0 0 40px 0; letter-spacing: 1px;">U.S. Domestic Flights · 2019–2023</h2>

<hr style="border: none; border-top: 1px solid #a8b2d833; margin: 0 auto 40px auto; width: 60%;">

<div style="display: flex; justify-content: center;">
<table style="color: #ccd6f6; font-size: 15px; border-collapse: collapse;">
  <tr>
    <td style="text-align: right; padding: 6px 16px; color: #a8b2d8;">📋 Phase</td>
    <td style="text-align: left; padding: 6px 16px; font-weight: 600;">2 — Machine Learning Models & Model Evaluation</td>
  </tr>
  <tr>
    <td style="text-align: right; padding: 6px 16px; color: #a8b2d8;">🎓 Course</td>
    <td style="text-align: left; padding: 6px 16px; font-weight: 600;">Ciência de Dados</td>
  </tr>
  <tr>
    <td style="text-align: right; padding: 6px 16px; color: #a8b2d8;">👨‍🎓 Student</td>
    <td style="text-align: left; padding: 6px 16px; font-weight: 600;">Micael Ribeiro</td>
  </tr>
  <tr>
    <td style="text-align: right; padding: 6px 16px; color: #a8b2d8;">👨‍🏫 Professor</td>
    <td style="text-align: left; padding: 6px 16px; font-weight: 600;">Fábio Mendonça</td>
  </tr>
  <tr>
    <td style="text-align: right; padding: 6px 16px; color: #a8b2d8;">📅 Date</td>
    <td style="text-align: left; padding: 6px 16px; font-weight: 600;">19-05-2026</td>
  </tr>
</table>
</div>

<hr style="border: none; border-top: 1px solid #a8b2d833; margin: 40px auto 0 auto; width: 60%;">

<p style="color: #a8b2d8; font-size: 12px; margin: 20px 0 0 0; letter-spacing: 1px; text-align: center;">U.S. Department of Transportation · On-Time Performance Reporting System</p>

</div>

# 🤖 Phase 4: Model Building

## 🎯 Objective

The objective of Phase 4 is to implement and evaluate multiple machine learning approaches for predicting and analyzing flight delays. Following preprocessing, exploratory analysis, hypothesis testing, feature engineering, and model selection, several predictive and unsupervised learning models are developed to assess their ability to capture operational delay patterns within the dataset.

The implemented approaches include:
- Distance-based learning methods
- Traditional supervised learning models
- Ensemble learning models using bagging and boosting
- A deep learning neural network implemented with PyTorch
- Clustering algorithms for unsupervised pattern discovery

All predictive models use only information available before departure in order to avoid data leakage and simulate realistic flight delay prediction conditions.

---

## 📌 Modeling Strategy

This phase includes three main machine learning tasks:

### 📈 Regression
The regression task predicts the numerical arrival delay duration in minutes using `ARR_DELAY` as the target variable.

### 🏷️ Classification
The classification task transforms `ARR_DELAY` into three operational categories:

- ✅ On-time: arrival delay < 15 minutes
- ⏱️ Short delay: 15–30 minutes
- ⏳ Long delay: > 30 minutes

### 🔍 Clustering
The clustering task identifies hidden operational patterns within airlines, airports, and flight routes without using predefined labels.

---

## ⚙️ Data Preparation for Modeling

Before training the models, the dataset was prepared using the following preprocessing steps:

- Removed leakage-related columns containing post-flight information
- Removed cancelled and diverted flights from prediction tasks
- Filled remaining missing numeric values using median imputation
- Incorporated engineered features created during Phase 3
- Selected numeric variables for model training
- Applied feature scaling where necessary
- Split the dataset into training and testing subsets using a fixed random state for reproducibility

The target variable for regression was `ARR_DELAY`.
For classification tasks, `ARR_DELAY` was transformed into categorical delay classes based on the project thresholds.

---


## 📂 Loading the Final Modeling Dataset

The preprocessing, leakage removal, scaling, and feature engineering steps were completed in the previous notebook.
For the modeling phase, the final processed dataset containing all engineered features is loaded directly from the saved CSV file.

This dataset includes:
- Cleaned flight records
- Scaled numerical variables
- Encoded categorical variables
- Engineered temporal and interaction features
- Leakage-free predictors suitable for modeling

---

In [1]:
import pandas as pd

df = pd.read_csv("../ProjectDatasets/flights_clean_scaled_featured.csv")
df.head()

,CRS_DEP_TIME,CRS_ARR_TIME,ARR_DELAY,CRS_ELAPSED_TIME,DISTANCE,CRS_ELAPSED_TIME_std,DISTANCE_std,CRS_DEP_TIME_std,CRS_ARR_TIME_std,CRS_ELAPSED_TIME_minmax,...,DEP_HOUR,dep_time_bin,flight_duration_bin,distance_bin,elapsed_x_distance,dep_hour_x_elapsed,DAY_OF_WEEK,MONTH,IS_WEEKEND,IS_RUSH_HOUR
0,1155,1501,0.0,186.0,1065.0,1.035741,0.431214,-0.352292,0.021842,0.602151,...,11,Morning,Medium-Long,Medium-Long,198090.0,2046.0,2,1,0,0
1,2120,2315,0.0,235.0,1399.0,1.957924,0.997970,1.634981,1.614626,0.777778,...,21,Evening/Night,Medium-Long,Medium-Long,328765.0,4935.0,5,11,1,0
2,954,1252,0.0,118.0,680.0,-0.244024,-0.222083,-0.766222,-0.465385,0.358423,...,9,Morning,Medium-Short,Medium-Short,80240.0,1062.0,4,7,0,0
3,1609,1829,24.0,260.0,1589.0,2.428426,1.320376,0.582653,0.663652,0.867384,...,16,Afternoon,Long,Medium-Long,413140.0,4160.0,0,3,0,1
4,1840,2041,0.0,181.0,985.0,0.941640,0.295464,1.058363,1.078480,0.584229,...,18,Evening/Night,Medium-Long,Medium-Short,178285.0,3258.0,6,2,1,1


# 🧠 4.1 k-Nearest Neighbors (kNN) From Scratch

## 📌 Objective

The first machine learning model implemented in this phase was a custom k-Nearest Neighbors (kNN) algorithm developed entirely from scratch using NumPy arrays. The purpose of this implementation was to understand the internal mechanics of distance-based learning instead of relying on the built-in `scikit-learn` implementation.

The kNN algorithm was applied to both:
- 📈 Regression: predicting the numerical arrival delay (`ARR_DELAY`)
- 🏷️ Classification: predicting delay categories (on-time, short delay, long delay)

---

## ⚙️ kNN Algorithm Overview

The k-Nearest Neighbors algorithm is a non-parametric and instance-based learning method. Instead of learning explicit model parameters during training, the algorithm stores the training data and makes predictions based on the similarity between observations.

For each test sample:
1. The Euclidean distance to all training samples is calculated
2. The `k` closest neighbors are selected
3. Predictions are generated:
   - Regression → average of neighbor target values
   - Classification → majority vote among neighbor classes

The Euclidean distance metric used in this implementation is:

```math
d(x,y)=\sqrt{\sum_{i=1}^{n}(x_i-y_i)^2}
```

---

## 📊 Evaluation Strategy

Multiple values of `k` were tested in order to evaluate how neighborhood size affects model performance. The tested values were configured through the YAML configuration file, allowing flexible experimentation without modifying the source code directly.

The following `k` values were evaluated:

```yaml
k_values: [13, 15, 17, 19, 21, 23, 25, 27, 29]
```

Higher `k` values generally produce smoother predictions and reduce sensitivity to noise, while lower values can capture local patterns more aggressively but may increase overfitting. The selected range was intended to balance stability and predictive performance across both regression and classification tasks.

The following evaluation metrics were used:

### Regression Metrics
- Mean Absolute Error (MAE)
- Root Mean Squared Error (RMSE)
- R² Score

### Classification Metrics
- Accuracy
- Precision
- Recall
- F1-score
- Confusion Matrix

---

In [2]:
import yaml
import pandas as pd
import contextlib
import io

from ProjectCodes.PythonProject.src.kNN import KNNRunner

with open("config.yaml", "r") as f:
    config = yaml.safe_load(f)

knn_runner = KNNRunner(df, config=config)

f = io.StringIO()

with contextlib.redirect_stdout(f):
    regression_results = knn_runner.run_regression()
    classification_results = knn_runner.run_classification()

regression_df = pd.DataFrame(regression_results)
classification_df = pd.DataFrame(classification_results)

print("kNN Regression Results")
display(regression_df)

print("kNN Classification Results")
display(classification_df)

kNN Regression Results


,Model,k,MAE,RMSE,R2
0,kNN Regression,13,20.449977,46.826638,-0.076763
1,kNN Regression,15,20.419427,46.642807,-0.068325
2,kNN Regression,17,20.339335,46.446247,-0.059340
3,kNN Regression,19,20.281000,46.309028,-0.053090
4,kNN Regression,21,20.271943,46.186429,-0.047521
5,kNN Regression,23,20.242678,46.056738,-0.041647
6,kNN Regression,25,20.207752,45.965645,-0.037530
7,kNN Regression,27,20.158430,45.898653,-0.034508
8,kNN Regression,29,20.156231,45.816147,-0.030792


kNN Classification Results


,Model,k,Accuracy,Macro Precision,Macro Recall,Macro F1,Weighted Precision,Weighted Recall,Weighted F1
0,kNN Classification,3,0.473479,0.477537,0.473052,0.473536,0.478629,0.473479,0.474328
1,kNN Classification,5,0.442336,0.446213,0.441807,0.439339,0.446718,0.442336,0.439927
2,kNN Classification,7,0.427737,0.427413,0.426986,0.425667,0.427895,0.427737,0.426300
3,kNN Classification,9,0.420925,0.422995,0.421013,0.420448,0.423405,0.420925,0.420639
4,kNN Classification,11,0.424331,0.428586,0.424226,0.423828,0.428864,0.424331,0.424064
5,kNN Classification,13,0.423844,0.425901,0.423563,0.423045,0.426173,0.423844,0.423347
6,kNN Classification,15,0.402433,0.405085,0.402512,0.401590,0.405320,0.402433,0.401709


## 💬 kNN Results Discussion

The regression results showed a gradual improvement as the value of `k` increased. Larger neighborhood sizes produced lower MAE and RMSE values, indicating that averaging over more neighboring samples reduced sensitivity to noise and outliers in the dataset.

For the regression task, higher values of `k` were used (`13–29`) because larger neighborhood sizes generally produce more stable continuous predictions by reducing local variance.

The best regression performance was achieved with `k=29`, producing:

* MAE ≈ 20.16
* RMSE ≈ 45.82
* R² ≈ -0.031

Although prediction error improved slightly with larger values of `k`, all regression models still produced negative R² scores. This indicates that the custom kNN regressor performed worse than a simple baseline predictor using the mean arrival delay. These results suggest that the flight delay dataset contains complex nonlinear relationships and substantial variability that are difficult to capture using a basic distance-based approach.

For the classification task, a separate range of smaller `k` values (`3–15`) was tested, since classification problems typically benefit from smaller neighborhood sizes that preserve local class boundaries more effectively.

The best classification performance was achieved with `k=3`, producing:

* Accuracy ≈ 47.3%
* Macro F1 ≈ 47.4%

Classification performance gradually decreased as the value of `k` increased, suggesting that excessive smoothing between neighborhoods reduced the model’s ability to distinguish between delay categories.

These results are also consistent with guidance provided by the professor on Discord, where classification accuracies around 50% were discussed as realistic for this dataset and problem formulation. The obtained results therefore align with the expected difficulty of predicting flight delay categories using only pre-departure information.

Overall, the custom kNN implementation successfully demonstrated the mechanics of instance-based learning, Euclidean distance computation, regression, and multi-class classification using NumPy only. However, the model still struggled with the scale, dimensionality, and complexity of the flight delay dataset, motivating the use of more advanced supervised, ensemble, and deep learning methods in later sections.


# 🌳 4.2 Supervised Learning Models

## 📌 Objective

Following the custom kNN implementation, additional supervised learning algorithms were evaluated using the `scikit-learn` framework. These models were selected to compare different learning strategies and assess their ability to capture nonlinear relationships within the flight delay dataset.

The supervised learning phase included:
- 🌲 Decision Tree Regressor
- 📏 Support Vector Regressor (SVR)

Both models were applied to the regression task of predicting numerical arrival delay duration (`ARR_DELAY`).

---

## ⚙️ Model Overview

### 🌲 Decision Tree Regressor

Decision Trees partition the feature space into smaller regions by recursively splitting the data according to feature thresholds. The model learns decision rules that minimize prediction error while capturing nonlinear relationships and feature interactions.

Advantages:
- Interpretable structure
- Handles nonlinear patterns
- Does not require feature scaling

Limitations:
- Sensitive to overfitting
- High variance on noisy datasets

---

### 📏 Support Vector Regressor (SVR)

Support Vector Regression attempts to fit a function within a specified error margin while maximizing generalization performance. SVR is capable of modeling nonlinear relationships using kernel transformations.

Advantages:
- Effective for nonlinear regression
- Strong theoretical generalization properties
- Robust to high-dimensional spaces

Limitations:
- Computationally expensive on large datasets
- Sensitive to parameter tuning and feature scaling

---

## 📊 Evaluation Strategy

The supervised learning models were evaluated using standard regression metrics:

- Mean Absolute Error (MAE)
- Root Mean Squared Error (RMSE)
- R² Score

Model hyperparameters were configured through the YAML configuration file, allowing experimentation without directly modifying the source code.

---

In [10]:
import yaml
import contextlib
import io

from ProjectCodes.PythonProject.src.supervised_learning import SupervisedLearningRunner

with open("config.yaml", "r") as f:
    config = yaml.safe_load(f)

supervised_runner = SupervisedLearningRunner(df, config=config)

f = io.StringIO()

with contextlib.redirect_stdout(f):
    supervised_results = supervised_runner.run_all()

display(supervised_results)

,Model,MAE,RMSE,R2
0,Decision Tree Regressor,19.276482,47.772187,0.014408
1,Support Vector Regressor,13.541305,51.177574,-0.055482


## 💬 Supervised Learning Results Discussion

The supervised learning models produced mixed results when applied to the flight delay regression task.

The **Decision Tree Regressor** achieved:
- MAE ≈ 19.28
- RMSE ≈ 47.77
- R² ≈ 0.014

Among the two supervised models, the Decision Tree produced the best overall performance. The model slightly outperformed the baseline predictor, as indicated by the small positive R² score. This suggests that the tree-based structure was capable of capturing some nonlinear relationships and interactions between scheduling, airline, and route-related variables.

Decision Trees also provided additional interpretability through feature importance analysis and visual tree structures. However, the relatively low R² score indicates that flight delays remain difficult to predict accurately due to the high variability and operational complexity of the dataset.

---

The **Support Vector Regressor (SVR)** achieved:
- MAE ≈ 13.54
- RMSE ≈ 51.18
- R² ≈ -0.055

The SVR model produced the lowest MAE among the evaluated supervised models, indicating that average prediction error was reduced for many samples. However, the negative R² score suggests that the model still struggled to explain the overall variance in arrival delays and performed worse than a simple mean-based baseline predictor.

This behavior may be related to:
- the large scale of the dataset
- the complexity of flight delay patterns
- sensitivity of SVR to parameter tuning
- limitations of the selected kernel and hyperparameters

Although SVR can model nonlinear relationships effectively, it is computationally expensive on large datasets and may require extensive tuning to achieve stronger performance.

---

Overall, the supervised learning models demonstrated improvements over the custom kNN implementation, particularly in terms of prediction error reduction. The Decision Tree model provided the strongest balance between interpretability and predictive capability, while SVR highlighted the challenges of applying margin-based methods to noisy operational datasets.

These results motivate the exploration of ensemble learning methods in the next phase, where combining multiple models may improve generalization performance and robustness.

---

# 🌲 4.3 Ensemble Learning Models

## 📌 Objective

Following the supervised learning models, ensemble learning techniques were implemented to improve predictive performance and model robustness. Ensemble methods combine multiple weaker learners in order to produce more stable and accurate predictions than individual models alone.

The ensemble learning phase included:
- 🌳 Random Forest Regressor
- 🚀 Gradient Boosting Regressor

Both models were applied to the regression task of predicting numerical arrival delay duration (`ARR_DELAY`).

---

## ⚙️ Model Overview

### 🌳 Random Forest Regressor

Random Forest is a bagging-based ensemble method that combines predictions from multiple Decision Trees trained on random subsets of the data and features. The final prediction is obtained by averaging the outputs of all trees.

Advantages:
- Reduces overfitting compared to a single Decision Tree
- Handles nonlinear relationships effectively
- Robust to noise and outliers
- Provides feature importance analysis

Limitations:
- Less interpretable than a single tree
- Higher computational cost

---

### 🚀 Gradient Boosting Regressor

Gradient Boosting builds trees sequentially, where each new tree attempts to correct the prediction errors of the previous ensemble. This boosting strategy allows the model to gradually improve predictive accuracy.

Advantages:
- Strong predictive performance
- Captures complex nonlinear patterns
- Effective on structured/tabular datasets

Limitations:
- Sensitive to hyperparameter tuning
- More prone to overfitting if boosting is too aggressive
- Slower training compared to simpler models

---

## 📊 Evaluation Strategy

The ensemble learning models were evaluated using the following regression metrics:

- Mean Absolute Error (MAE)
- Root Mean Squared Error (RMSE)
- R² Score

Model hyperparameters such as the number of estimators, tree depth, and learning rate were configured through the YAML configuration file, allowing flexible experimentation without directly modifying the implementation code.

---

In [11]:
import yaml
import contextlib
import io

from ProjectCodes.PythonProject.src.ensemble_learning import EnsembleLearningRunner

with open("config.yaml", "r") as f:
    config = yaml.safe_load(f)

ensemble_runner = EnsembleLearningRunner(df, config=config)

f = io.StringIO()

with contextlib.redirect_stdout(f):
    ensemble_results = ensemble_runner.run_all()

display(ensemble_results)

,Model,MAE,RMSE,R2
0,Random Forest Regressor (Bagging),19.168870,47.635635,0.020035
1,Gradient Boosting Regressor (Boosting),19.297558,47.716276,0.016714


## 💬 Ensemble Learning Results Discussion

The ensemble models produced the strongest regression results so far.

The **Random Forest Regressor** achieved:
- MAE ≈ 19.17
- RMSE ≈ 47.64
- R² ≈ 0.020

The **Gradient Boosting Regressor** achieved:
- MAE ≈ 19.30
- RMSE ≈ 47.72
- R² ≈ 0.017

Random Forest performed slightly better overall, with the lowest MAE and RMSE and the highest R² score. This suggests that bagging multiple decision trees improved stability compared to a single Decision Tree model.

Gradient Boosting also performed well, but it did not outperform Random Forest in this run. This may indicate that the selected boosting hyperparameters were conservative or that further tuning would be needed to fully benefit from sequential error correction.

Overall, ensemble learning improved model robustness compared with simpler methods such as kNN and performed similarly or slightly better than the standalone Decision Tree. However, the low R² values show that arrival delay prediction remains challenging due to high variability and operational noise in the dataset.

---

# 🧠 4.4 Deep Learning Model

## 📌 Objective

Following the traditional supervised and ensemble learning approaches, a deep learning model was implemented to evaluate whether neural networks could better capture the complex nonlinear relationships present in the flight delay dataset.

The deep learning phase used a Multi-Layer Perceptron (MLP) implemented with PyTorch for the regression task of predicting numerical arrival delay duration (`ARR_DELAY`).

---

## ⚙️ Model Overview

### 🧠 Multi-Layer Perceptron (MLP)

A Multi-Layer Perceptron is a feedforward neural network composed of multiple fully connected layers. The network learns hierarchical feature representations by applying nonlinear activation functions between layers.

The implemented architecture included:
- Input layer matching the number of dataset features
- Hidden fully connected layers with nonlinear activations
- Output layer for regression prediction

The model was trained using backpropagation and gradient-based optimization.

Advantages:
- Capable of learning complex nonlinear relationships
- Flexible architecture
- Suitable for high-dimensional data

Limitations:
- Requires careful hyperparameter tuning
- Computationally expensive
- More difficult to interpret compared to tree-based models

---

## 📊 Evaluation Strategy

The deep learning model was evaluated using standard regression metrics:

- Mean Absolute Error (MAE)
- Root Mean Squared Error (RMSE)
- R² Score

Training configuration parameters such as:
- hidden layer size
- learning rate
- batch size
- number of epochs

were controlled through the YAML configuration file to allow flexible experimentation without modifying the implementation code directly.

---

In [12]:
import yaml
import pandas as pd
import contextlib
import io

from ProjectCodes.PythonProject.src.deep_learning import DeepLearningRunner

with open("config.yaml", "r") as f:
    config = yaml.safe_load(f)

deep_runner = DeepLearningRunner(df, config=config)

f = io.StringIO()

with contextlib.redirect_stdout(f):
    deep_results = deep_runner.run_all()

deep_results_df = pd.DataFrame([deep_results])

display(deep_results_df)

,Model,MAE,RMSE,R2
0,PyTorch Feed-Forward Neural Network,19.356643,51.918425,0.010949


## 💬 Deep Learning Results Discussion

The PyTorch Feed-Forward Neural Network achieved:

- MAE ≈ 19.36
- RMSE ≈ 51.92
- R² ≈ 0.011

The neural network achieved a small positive R² score, meaning it performed slightly better than a mean-based baseline predictor. However, it did not outperform the tree-based ensemble models.

This result is reasonable because the dataset is tabular and highly noisy. Deep learning models often require careful tuning, larger training time, and more optimized architectures to outperform traditional machine learning methods on structured tabular data.

The model still demonstrated the ability to learn some nonlinear relationships, but the relatively high RMSE shows that large delay cases remained difficult to predict accurately.

Overall, the deep learning model satisfied the project requirement and provided a useful comparison point. However, for this dataset, traditional tree-based models such as Random Forest and Gradient Boosting were more effective and easier to interpret.

---

# 🔍 4.5 Clustering Analysis

## 📌 Objective

The final model-building task was clustering, which was used to identify hidden operational patterns in the flight delay dataset. Unlike the previous models, clustering is an unsupervised learning approach, meaning that it does not use a target variable for prediction.

The goal of this section was to group flights based on operational characteristics such as delay behavior, distance, scheduled flight duration, departure time, and engineered features. These clusters can help reveal patterns related to airline punctuality, airport congestion, route behavior, and general operational performance.

---

## ⚙️ Clustering Models

Two clustering algorithms were applied:

### 🔵 KMeans Clustering

KMeans is a centroid-based clustering algorithm that partitions observations into a predefined number of clusters. The algorithm attempts to minimize the distance between each observation and the center of its assigned cluster.

In this project, multiple values of `k` were tested to evaluate whether meaningful cluster structures existed in the data.

Advantages:
- Simple and efficient
- Useful for identifying broad operational groupings
- Works well when clusters are relatively compact and separated

Limitations:
- Requires choosing the number of clusters
- Sensitive to feature scaling
- Assumes roughly spherical cluster shapes

---

### 🟣 DBSCAN Clustering

DBSCAN is a density-based clustering algorithm that groups observations based on local point density. Unlike KMeans, it does not require specifying the number of clusters beforehand and can identify noise points or outliers.

Advantages:
- Can detect irregular cluster shapes
- Identifies noise and anomaly-like observations
- Does not require a fixed number of clusters

Limitations:
- Sensitive to `eps` and `min_samples`
- Can struggle when data density varies strongly
- May classify many points as noise depending on parameter settings

---

## 📊 Evaluation Strategy

The clustering models were evaluated using:

- Inertia for KMeans
- Silhouette Score for cluster separation
- Cluster counts
- DBSCAN number of clusters
- DBSCAN noise ratio
- PCA-based visualizations of cluster assignments

The number of clusters and DBSCAN parameters were configured through the YAML configuration file, allowing experimentation without directly modifying the implementation code.

---

In [13]:
import yaml
import pandas as pd
import contextlib
import io

from ProjectCodes.PythonProject.src.clustering import ClusteringRunner

with open("config.yaml", "r") as f:
    config = yaml.safe_load(f)

clustering_runner = ClusteringRunner(df, config=config)

f = io.StringIO()

with contextlib.redirect_stdout(f):
    clustering_runner.run_all()

kmeans_results = pd.read_csv(config["output_dir_model_results"] + "/clustering_kmeans_results.csv")
dbscan_results = pd.read_csv(config["output_dir_model_results"] + "/clustering_dbscan_results.csv")
kmeans_counts = pd.read_csv(config["output_dir_model_results"] + "/clustering_kmeans_cluster_counts.csv")
dbscan_counts = pd.read_csv(config["output_dir_model_results"] + "/clustering_dbscan_cluster_counts.csv")

print("KMeans Results")
display(kmeans_results)

print("DBSCAN Results")
display(dbscan_results)

print("KMeans Cluster Counts")
display(kmeans_counts)

print("DBSCAN Cluster Counts")
display(dbscan_counts)

KMeans Results


,Algorithm,k,Inertia,Silhouette Score
0,KMeans,2,543234.272227,0.245651
1,KMeans,3,482659.303102,0.171146
2,KMeans,4,437290.080329,0.169659
3,KMeans,5,396102.522928,0.171077
4,KMeans,6,362645.668498,0.168542
5,KMeans,7,338328.558121,0.170150
6,KMeans,8,313882.957612,0.183573


DBSCAN Results


,Algorithm,eps,min_samples,Clusters Found,Noise Points,Noise Ratio
0,DBSCAN,0.8,15,7,2979,0.02979


KMeans Cluster Counts


,Cluster,Count
0,0,36529
1,1,63471


DBSCAN Cluster Counts


,Cluster,Count
0,-1,2979
1,0,92297
2,1,4618
3,2,39
4,3,24
5,4,17
6,5,15
7,6,11


## 💬 Clustering Results Discussion

The clustering analysis identified several operational groupings within the flight delay dataset using both centroid-based and density-based clustering methods.

---

### 🔵 KMeans Results

KMeans clustering was evaluated using values of `k` ranging from 2 to 8.

The highest silhouette score was achieved at:

- `k = 2`
- Silhouette Score ≈ 0.246

As the number of clusters increased, inertia continued to decrease, which is expected because additional clusters reduce within-cluster variance. However, the silhouette scores generally decreased or remained relatively low, suggesting that the dataset does not contain strongly separated cluster structures.

The final KMeans clustering produced two dominant operational groups:
- Cluster 0 → approximately 36,529 flights
- Cluster 1 → approximately 63,471 flights

These clusters likely represent broad operational behaviors such as:
- lower-delay vs higher-delay operational patterns
- short-haul vs long-haul flight structures
- different scheduling and traffic profiles

The relatively modest silhouette scores indicate that operational flight behavior overlaps significantly across flights, airlines, and routes.

---

### 🟣 DBSCAN Results

DBSCAN clustering produced:
- 7 detected clusters
- approximately 2.98% noise points

Most observations were concentrated in one large dense cluster:
- Cluster 0 → approximately 92,297 flights

Several much smaller clusters and anomaly-like groups were also detected. These likely correspond to unusual operational behaviors, rare delay patterns, or less common route and scheduling combinations.

The presence of noise points demonstrates one advantage of DBSCAN compared to KMeans: the ability to identify observations that do not belong clearly to any dense operational pattern.

---

### 📌 Overall Interpretation

The clustering analysis revealed that the flight delay dataset contains broad operational structures but relatively weak cluster separation overall. This is reasonable because flight operations are influenced by many overlapping factors such as:
- airline schedules
- airport congestion
- weather conditions
- route distance
- seasonal effects

Although the clusters were not strongly separated, the analysis still provided insight into underlying operational behaviors and demonstrated how unsupervised learning can be used to explore patterns in airline and airport performance without predefined labels.

---

# 📊 Phase 5: Model Comparison and Evaluation

## 📌 Objective

The objective of Phase 5 is to compare the performance of all implemented machine learning models, evaluate their strengths and weaknesses, summarize the main findings of the project, and identify possible directions for future improvements.

The evaluation focuses on:

* predictive performance
* model robustness
* interpretability
* computational complexity
* suitability for operational flight delay prediction

---

## 📈 Regression Model Comparison

| Model                               |   MAE |  RMSE |     R² |
| ----------------------------------- | ----: | ----: | -----: |
| kNN Regression                      | 20.16 | 45.82 | -0.031 |
| Decision Tree Regressor             | 19.28 | 47.77 |  0.014 |
| Support Vector Regressor            | 13.54 | 51.18 | -0.055 |
| Random Forest Regressor             | 19.17 | 47.64 |  0.020 |
| Gradient Boosting Regressor         | 19.30 | 47.72 |  0.017 |
| PyTorch Feed-Forward Neural Network | 19.36 | 51.92 |  0.011 |

---

## 🏷️ Classification Model Comparison

| Model              | Accuracy | Macro F1 | Observation                                                       |
| ------------------ | -------: | -------: | ----------------------------------------------------------------- |
| kNN Classification |    0.473 |    0.474 | Smaller k values preserved local class structure more effectively |

---

## 🔍 Clustering Model Comparison

| Algorithm          | Main Result                                                                       |
| ------------------ | --------------------------------------------------------------------------------- |
| KMeans             | Best silhouette score achieved at k=2                                             |
| DBSCAN             | Detected 7 clusters with approximately 3% noise points                            |
| Airline Clustering | Grouped airlines according to operational delay behavior and punctuality profiles |

---

# ⚖️ Strengths and Weaknesses of Each Approach

## 🧠 k-Nearest Neighbors (kNN)

### Strengths

* Simple and intuitive algorithm
* Easy to implement from scratch
* No explicit training phase required
* Demonstrated both regression and multi-class classification capabilities

### Weaknesses

* Computationally expensive during prediction
* Sensitive to feature scaling
* Struggled with high-dimensional data
* Performance decreased for larger neighborhood sizes in classification tasks

---

## 🌲 Decision Tree Regressor

### Strengths

* Interpretable model structure
* Captures nonlinear feature interactions
* Provides feature importance information

### Weaknesses

* Sensitive to overfitting
* Limited generalization performance
* Performance depends heavily on tree depth configuration

---

## 📏 Support Vector Regressor (SVR)

### Strengths

* Effective for nonlinear regression
* Strong theoretical generalization properties
* Produced the lowest MAE among supervised models

### Weaknesses

* Computationally expensive on large datasets
* Sensitive to parameter tuning
* Negative R² indicated poor variance explanation

---

## 🌳 Random Forest Regressor

### Strengths

* Strongest overall regression performance
* Robust to noise and overfitting
* Handles nonlinear relationships effectively

### Weaknesses

* Less interpretable than a single tree
* Higher computational cost
* Large ensemble models require more memory

---

## 🚀 Gradient Boosting Regressor

### Strengths

* Strong predictive capability
* Effective sequential error correction
* Handles complex patterns well

### Weaknesses

* Sensitive to hyperparameter tuning
* Slower training process
* Can overfit if boosting becomes too aggressive

---

## 🧠 Deep Learning Neural Network

### Strengths

* Learns complex nonlinear relationships
* Flexible architecture
* Suitable for high-dimensional data

### Weaknesses

* Difficult to interpret
* Requires significant tuning
* Did not outperform ensemble tree models on this tabular dataset

---

## 🔍 Clustering Methods

### Strengths

* Revealed operational flight behavior patterns
* Identified dense operational groups and anomalies
* Airline clustering aligned directly with operational performance analysis objectives
* Useful for exploratory analysis

### Weaknesses

* Weak cluster separation overall
* Sensitive to parameter selection
* Operational patterns overlapped significantly

---

# 📌 Overall Findings and Insights

The project demonstrated that predicting flight delays is a highly challenging task due to the noisy and operationally complex nature of airline systems.

Several important findings emerged from the analysis:

* Ensemble learning models achieved the strongest overall regression performance
* Random Forest produced the best balance between robustness and predictive capability
* Distance-based methods such as kNN struggled with high-dimensional operational data
* Smaller neighborhood sizes performed better for classification tasks
* Deep learning did not outperform traditional ensemble methods for this structured tabular dataset
* Clustering revealed broad operational groupings but relatively weak cluster separation

The classification experiments also showed that different k ranges were more suitable depending on the task formulation. Larger neighborhood sizes produced smoother and more stable regression predictions, while smaller neighborhood sizes better preserved local class boundaries for classification. The obtained classification accuracy was also consistent with values discussed by the course instructor on Discord, where results around 50% were considered realistic for this dataset using only pre-departure information.

The results also highlighted that airline delay prediction is influenced by many overlapping factors including scheduling, route structure, airport congestion, and operational variability.

Overall, the project successfully demonstrated the full machine learning workflow:

* preprocessing
* feature engineering
* hypothesis testing
* supervised learning
* ensemble learning
* deep learning
* clustering
* model evaluation and interpretation

---

# 🚀 Recommendations and Future Improvements

Several improvements could further enhance model performance and operational usefulness:

* Apply advanced hyperparameter optimization techniques
* Explore more advanced ensemble methods such as XGBoost or LightGBM
* Incorporate external weather and air traffic datasets
* Develop route-specific or airline-specific prediction models
* Experiment with alternative distance metrics and dimensionality reduction techniques for kNN
* Experiment with temporal deep learning architectures such as LSTMs or Transformers
* Build real-time deployment pipelines for operational delay forecasting
* Increase computational resources for larger-scale deep learning experiments

---
